### PR TITLE
【Hackathon 6th No.33】支持动态图流水并行设定多个损失函数，并返回多个 loss

### PR DESCRIPTION
**Pull request by DanGuge**
### PR Category

Distributed Strategy

### PR Types

Improvements

### Description

1. 在构建 PipelineLayer 时，对用户传入的 loss_fn 进行判断
  * 如果 loss_fn 不是 list，则将 loss_fn 用 [] 包裹
  * 如果 loss_fn 是 list，则直接设置
2. 在计算 loss 时，将计算和通信逻辑都调整为 list loss_fn 的逻辑
3. 在返回 loss 时，根据 len(losses) 的长度，判断用户需要返回 1 个 loss 结果或者多个 loss 结果

### CODE REVIEW

**Comment by DanGuge** [comment]
@MarioLulab 辛苦review一下～

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 1]
```diff
@@ -776,22 +776,30 @@ def _forward_step(self, input_tensor, micro_dataset, chunk_id=None):
...
-                with paddle.amp.auto_cast(enable=False):
-                    if self.accumulate_steps > 1 and not self._delay_scale_loss:
-                        output_tensor = output_tensor / self.accumulate_steps
-
-                    if self.total_loss is None:
-                        self.total_loss = paddle.zeros_like(output_tensor)
-                    self.total_loss += output_tensor.detach()
+                for idx, loss_fn in enumerate(self._layers._loss_fn):
+                    output_tensor = loss_fn(output_tensor, labels)
+                    assert isinstance(
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:809]: 在 len(_loss_fn) > 1 时，`output_tensor` 会被下一次循环覆盖掉，是否可以改成：

```python
loss_tensor = loss_fn(output_tensor, labels)
```

Comment (MarioLulab) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:809]: 当前的 `_forward_step` 不符合赛题的要求：「用户可以指定针对哪一个 loss 做 backward。默认下，只对返回的第 0 个 loss 做 backward」。照目前的写法，用户当前情况下没法控制针对哪一个 loss 做 backward。可以尝试在 `train_batch` 函数里新增入参，可让用户指明是对哪个 loss 做 backward

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 2]
```diff
@@ -851,22 +859,27 @@ def _broadcast_final_loss(self):
             assert (
                 self.total_loss is not None
             ), "train_batch() in last stage should obtain valid loss"
-            loss = (
-                self.total_loss.detach()
+            losses = [
+                self.total_loss[idx].detach()
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:893]: 这里的 `idx` 是指？

**Review by MarioLulab — COMMENTED** [review]
Nice Work~ 但还有一些问题，辛苦看下

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (DanGuge) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:809]: > 在 len(_loss_fn) > 1 时，`output_tensor` 会被下一次循环覆盖掉，是否可以改成：
> 
> ```python
> loss_tensor = loss_fn(output_tensor, labels)
> ```

这里的output_tensor只作为一个临时的中间变量，output会被添加到self.total_loss中，此处确实没有考虑到对指定的loss进行backward计算这个问题，目前这样实现是默认用的最后一个loss_fn来做的backward

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 2 above)
Comment (DanGuge) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:893]: 这里的idx在866行，如果loss_fn很多的话，total_loss中也会存有对应数量的由不同loss_fn计算得到的loss，此处是对所有loss_fn计算的loss进行聚合，在有多个loss_fn的时候，会返回一个list，存有[loss_0, loss_1, ...]

Comment (DanGuge) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:893]: 这里想问个问题，如果用户指定了loss_fn后，在train_batch/eval_batch中需要返回多个loss还是只返回用户指定的loss_fn对应的loss？
任务中是这么表述的：支持动态图流水线并行组网时指定多个损失函数，并返回多个 loss，用户可以指定针对哪一个 loss 做 backward
我目前的实现是如果只有一个loss_fn，只会返回一个loss，如果有多个loss_fn，会返回一个loss list

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (MarioLulab) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:809]: output_tensor 确实是作为一个临时的中间变量，但是我们期望达到的效果是不同的 loss_fn 之间是分开对前向得到的 `output_tensor` 与 `label` 求损失函数，而不是当前所写的串行的方式。题目期望效果如下：
image

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 2 above)
Comment (MarioLulab) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:893]: > 这里想问个问题，如果用户指定了loss_fn后，在train_batch/eval_batch中需要返回多个loss还是只返回用户指定的loss_fn对应的loss？ 任务中是这么表述的：支持动态图流水线并行组网时指定多个损失函数，并返回多个 loss，用户可以指定针对哪一个 loss 做 backward 我目前的实现是如果只有一个loss_fn，只会返回一个loss，如果有多个loss_fn，会返回一个loss list

如果用户指定对特定的 loss_fn 的计算结果进行 backward，在train_batch/eval_batch中也需要返回多个loss，即当 `loss_fn` 是一个 List 时，train_batch/eval_batch 应该返回 List[Tensor] or Tuple[Tensor]；当 `len(loss_fn) == 1`时，只需要返回 Tensor 即可（这样可以和原先的预期行为对齐）。

你目前的理解是对的。这个任务的背景其实是为了让用户使用 pp 时能够定义并返回多个 loss，方便 debug，所以需要返回多个 loss

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (DanGuge) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:809]: ok👌，之前实现确实有bug，fixed～

File: test/collective/fleet/hybrid_parallel_pp_multiple_losses.py
Comment (MarioLulab) [inline test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: 1. 目前该单测只是测试了2卡的 non-interleaved 模式的 pp。还需要添加 interleaved 模式的 pp 的单测。interleaved 模式的 pp 开启方法是：在派生 `PipelineLayer` 时，指定 `num_virtual_pipeline_stages` 参数，并且需满足 pp_degree > 2

2. 最好再增加一个多卡与单卡的逐位对齐单测，可以参考：https://github.com/PaddlePaddle/Paddle/pull/62563/files#diff-b5f15f3b078428d426f6ddcc1b7c57a67581366ecff2b3b754d9b41bb9484e8fR367-R448
所谓「逐位对齐」的含义是：在每一个 step 中，单卡模型计算得到的 loss 与多卡分布式模型计算得到的 loss 的误差保持在一定阈值内

**Review by MarioLulab — CHANGES_REQUESTED** [review]
还需要增加一些单测～

File: test/collective/fleet/hybrid_parallel_pp_multiple_losses.py
Comment (DanGuge) [reply test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: 单测添加了，辛苦review～

Comment (MarioLulab) [reply test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: ci-coverage 里提示代码覆盖率没过，辛苦看下是不是还需要添加一些单测或者单测格式不对导致有些代码没运行？

Comment (DanGuge) [reply test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: image

image

这个看起来不是我的单测，错误原因应该是很细微的精度上的问题，这种需要我来添加tolerance吗？

Comment (DanGuge) [reply test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: 还有一个error，我看https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/10427910/job/25821459里面也报错了，但是检查通过了
image
对应PR：https://github.com/PaddlePaddle/Paddle/pull/63222

Comment (MarioLulab) [reply test/collective/fleet/hybrid_parallel_pp_multiple_losses.py:1]: 当前 pr 的 ci-coverage 没过不是因为单测报错，而是因为「代码覆盖率」没过。
image

image

关于代码覆盖率的说明可以看：https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/git_guides/paddle_ci_manual_cn.html#pr-ci-coverage

**Comment by DanGuge** [comment]
@MarioLulab 辛苦review一下～

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 3]
```diff
@@ -686,8 +689,17 @@ def _wrap_data(self, data):
...
+    def train_batch(
+        self, data, optimizer, lr_scheduler=None, scaler=None, loss_fn_idx=0
+    ):
         data = self._prepare_training(data, optimizer, lr_scheduler)
+
+        # set loss function index
+        assert loss_fn_idx in range(
+            len(self._layers._loss_fn)
+        ), "loss_fn_idx should be in range of loss_fn"
+        self.loss_fn_idx = loss_fn_idx
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: 这里可以再加一个 assert ，断言 self._layers._loss_fn[loss_fn_idx] not None

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 4]
```diff
@@ -710,6 +722,12 @@ def eval_batch(self, data, compute_loss=False):
         # store total loss of entire batch
         self.total_loss = None
 
+        # set loss function index
+        assert loss_fn_idx in range(
+            len(self._layers._loss_fn)
+        ), "loss_fn_idx should be in range of loss_fn"
+        self.loss_fn_idx = loss_fn_idx
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: 同上

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 5]
```diff
@@ -780,29 +798,43 @@ def _forward_step(self, input_tensor, micro_dataset, chunk_id=None):
             # train calculate loss for train
             if self._compute_loss:
                 assert (
-                    self._layers._loss_fn is not None
+                    self._layers._loss_fn[0] is not None
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]:
```suggestion
                    self._layers._loss_fn[self.loss_fn_idx] is not None
```

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 6]
```diff
@@ -855,22 +887,27 @@ def _broadcast_final_loss(self):
...
-            )
+                else self.total_loss[idx] / self.accumulate_steps
+                for idx in range(len(self._layers._loss_fn))
+            ]
             is_fp32 = (
                 paddle.full([], 1, 'int64')
-                if loss.dtype == paddle.float32
+                if losses[0].dtype == paddle.float32
                 else paddle.full([], 0, 'int64')
             )
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: 此处对 `is_fp32` broadcast 也要对应处理为 list，因为可能不是 losses 里所有的 loss 都是 fp32 的

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 7]
```diff
@@ -1708,8 +1747,17 @@ def _process_bwd_buffer(step_id, tensor):
...
+    def train_batch(
+        self, data, optimizer, lr_scheduler=None, scaler=None, loss_fn_idx=0
+    ):
         data = self._prepare_training(data, optimizer, lr_scheduler)
+
+        # set loss function index
+        assert loss_fn_idx in range(
+            len(self._layers._loss_fn)
+        ), "loss_fn_idx should be in range of loss_fn"
+        self.loss_fn_idx = loss_fn_idx
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: 这里可以再加一个 assert ，断言 self._layers._loss_fn[loss_fn_idx] not None

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 8]
```diff
@@ -1719,13 +1767,19 @@ def train_batch(self, data, optimizer, lr_scheduler=None, scaler=None):
...
         self.set_virtual_pipeline_rank(0)
 
         self._layers.eval()
         self._compute_loss = compute_loss
 
+        # set loss function index
+        assert loss_fn_idx in range(
+            len(self._layers._loss_fn)
+        ), "loss_fn_idx should be in range of loss_fn"
+        self.loss_fn_idx = loss_fn_idx
```
Comment (MarioLulab) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: 同上

File: test/collective/fleet/hybrid_parallel_pp_multiple_losses_alignment.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+            x1 = paddle.to_tensor(x1_data)
+            x2 = paddle.to_tensor(x2_data)
+            y1 = paddle.to_tensor(y1_data)
+
+            x1.stop_gradient = True
+            x2.stop_gradient = True
+            y1.stop_gradient = True
+
+            loss_a = model_a(x1, x2, y1)
```
Comment (MarioLulab) [inline test/collective/fleet/hybrid_parallel_pp_multiple_losses_alignment.py:233]: 单卡模型的 loss 数量也要和多卡的对齐，返回 MSE, LossNet(), L1 三种 loss

**Review by MarioLulab — COMMENTED** [review]
nice work ～
我提了一些小修改，辛苦看下

**Comment by DanGuge** [comment]
> nice work ～ 我提了一些小修改，辛苦看下

@MarioLulab 修改好了，辛苦review～

**Review by MarioLulab — APPROVED** [review]
LGTM

**Comment by DanGuge** [comment]
@luotao1 辛苦review一下～

**Comment by luotao1** [comment]
辛苦 @ForFishes @GuoxiaWang 审核下

**Comment by DanGuge** [comment]
@GuoxiaWang @ForFishes 辛苦review一下～

**Comment by DanGuge** [comment]
@luotao1 辛苦review一下～

**Review by ForFishes — APPROVED** [review]
LGTM
